### PR TITLE
[NPUW] Support for qwen3-vl models

### PR DESCRIPTION
**Pull request by guweixin**
### Details:
  - Add Intel NPU runtime support for Qwen3-VL vision-language models in the NPUW plugin
  - Introduce ReplaceNonZero graph transformation to replace NonZero(Parameter["visual_pos_masks"]) with a pre-sized Parameter, enabling
  static compilation on NPU; transformation is scoped to visual_pos_masks parameters only to avoid side-effects on unrelated models
  - Add copy_deepstack_visual_embeds_to_right() helper to right-pad visual embeddings into a zero-padded static buffer
  - Add set_visual_pos_masks_input() helper to write NonZero-encoded indices or raw boolean mask depending on model input type
  - Extend prefill/generate inference paths to propagate visual_pos_masks and deepstack_visual_embeds
  - Read visual_pos_masks and deepstack_visual_embeds as optional ports in infer() — absent on non-VLM models — consistent with the
  token_type_ids pattern
  - Fix deepstack_visual_embeds static shape from {3, input_size, emb} to {1, input_size, emb} (batch=3 was a copy-paste error from
  position_ids)
  - Fix visual_pos_masks reshape: remove incorrect Whisper kvcache override that changed shape to {1, kvcache_size+1}
  - Replace deprecated get_any_name() with get_names() in reshape_to_static.cpp
  - Remove #ifdef NPU_PLUGIN_DEVELOPER_BUILD guard from NPUW dump options
  - Add unit tests: replace_nonzero_test.cpp (3 cases) and new cases in reshape_to_static_test.cpp (2 cases)
 - *...*
### Tickets:
 - *N/A*

### AI Assistance:
 - *AI assistance used: yes*
 - *Claude was used to identify review issues, implement fixes, and generate unit tests. Human validation performed: code review of all
  diffs, build verification on Windows with VS2022, manual inspection of test logic against existing repo patterns.*

### CODE REVIEW

**Comment by guweixin** [comment]
Hi, I'm an Intel employee(weixin.gu@intel.com). Could you please add my GitHub account (guweixin) to the openvinotoolkit organization, and trigger CI for this PR and review this PR? This adds Qwen3-VL NPU support via NPUW. Thanks!  @dmatveev

**Comment by guweixin** [comment]
build_jenkins

File: src/plugins/intel_npu/src/plugin/npuw/base_sync_infer_request.cpp [hunk 1]
```diff
@@ -200,8 +200,12 @@ ov::SoPtr<ov::ITensor> ov::npuw::IBaseInferRequest::get_tensor(const ov::Output<
         }
     }
 
-    NPUW_ASSERT(false);
-    return {};
+    std::string requested_port_name{"<unnamed>"};
+    try {
+        requested_port_name = port.get_any_name();
```
Comment (rkazants) [inline src/plugins/intel_npu/src/plugin/npuw/base_sync_infer_request.cpp:205]: `get_any_name()` is not correct to use because this is deprecated function

**Review by rkazants — CHANGES_REQUESTED** [review]
Please provide proper PR description, tests and avoid usage of deprecated stuff.

**Comment by guweixin** [comment]
@rkazants  Thanks for the review. I've addressed all the feedback in the latest commit (e3cda3c) 
Could you please approve the CI run? Thanks!

**Comment by guweixin** [comment]
@rkazants Thanks for the review. I've addressed all the feedback. Could you please approve the CI run? Thanks!

**Comment by dmatveev** [comment]
@AlexanderKalistratov please provide your review here.

File: src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/replace_deepstack_scatter_with_add.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        return false;
+    }
+
+    // The visual_pos_masks input now only feeds the dead gather/scatter cluster; drop it.
+    for (const auto& param : model->get_parameters()) {
+        if (is_named(param, "visual_pos_masks") && !is_reachable_from_outputs(model, param.get())) {
+            model->remove_parameter(param);
+            break;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/replace_deepstack_scatter_with_add.cpp:176]: breaks after remove, no fix needed

**Comment by AlexanderKalistratov** [comment]
build_jenkins

**Comment by AlexanderKalistratov** [comment]
# The original pattern
Runs **inside the model**, once per DeepStack level `L ∈ {0,1,2}` (after LM blocks 0/1/2). `N` = number of visual tokens (data-dependent), `S` = seq length, `D` = 2560.

```
Inputs to the LM graph:
    hidden                  : f32 [1, S, D]              (running hidden state)
    deepstack_visual_embeds : f32 [num_layers=3, N, D]   (read-only)
    visual_pos_masks        : bool [1, S]                (true at visual-token positions)

  visual_pos_masks [1,S] ── NonZero ──► [2, N] ── Transpose ──► idx [N, 2]
                                        (row0=batch=0, row1=seq pos)   (dynamic N!)

  For each level L (after block L):

        hidden [1,S,D] ──GatherND(idx)──►  hidden_vis [N, D] ─┐
                                                              ├─► Add ─► sum [N, D]
  deepstack[3,N,D] ─Gather(select, axis0, idx=L)─► ds_L [N,D]─┘           |
                                                                          |
                          ScatterNDUpdate(hidden, idx, sum) ◄─────────────┘
                                     │
                                     ▼
                          hidden' [1, S, D]      (only the N visual rows changed)
     (the Scatter is wrapped in a SliceAssign identity: Reshape → ScatterNDUpdate → Reshape)
```

Placement happens **in the graph** (`GatherND`/`ScatterNDUpdate` driven by
`NonZero(visual_pos_masks)`), and `N` is **dynamic** → not compilable to a static NPU
pipeline.

# Replacing original pattern with:
Matches the whole `ScatterNDUpdate(hidden, pos, Add(GatherND(hidden,pos), Gather(deepstack,L)))` cluster (+ the SliceAssign identity wrapper) and rewrites it to a plain residual add. The `Gather(select, L)` is kept; the deepstack param is now `[num_layers, S, D]` so it yields `[S, D]` per level.

```
Inputs to the (transformed) LM graph:
    hidden           : f32 [1, S, D]
    deepstack_static : f32 [num_layers=3, S, D]   (host already scattered)
    visual_pos_masks : REMOVED  (no longer used by the graph → parameter dropped)

  For each level L (after block L):

    deepstack_static[3,S,D] ─►Gather(select, axis0, idx=L)─► ds_L [S, D]
                                                             │  (broadcast over batch)
        hidden [1,S,D] ───────────► Add ◄────────────────────┘
                                     │
                                     ▼
                          hidden' [1, S, D]
```

File: src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/replace_deepstack_scatter_with_add.cpp [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+#include "openvino/pass/graph_rewrite.hpp"
+#include "openvino/pass/pattern/matcher.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "openvino/pass/validate.hpp"
+
+namespace opp = ov::pass::pattern;
+
+namespace {
+
+bool is_named(const std::shared_ptr<ov::Node>& node, const std::string& needle) {
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/replace_deepstack_scatter_with_add.cpp:27]: Do we want to preserve `needle` here?

File: src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/replace_deepstack_scatter_with_add.cpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                        cluster_out = reshape_out->output(0);
+                    }
+                }
+            }
+
+            // hidden + Gather(deepstack, L). The host pre-scatters the deepstack values into their
+            // visual-token positions (zeros elsewhere), so a plain broadcasted add reproduces the
+            // original per-position injection.
+            auto residual = std::make_shared<ov::op::v1::Add>(base, select_node->output(0));
+            residual->set_friendly_name(select_node->get_friendly_name() + "/deepstack_residual_add");
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/replace_deepstack_scatter_with_add.cpp]: Should we prefix the name with `npuw`?

File: src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/reshape_to_static.cpp [hunk 5]
```diff
@@ -32,11 +32,36 @@ void reshape_to_static(std::shared_ptr<ov::Model> model,
             NPUW_ASSERT(input.get_partial_shape().size() == 3u);
             NPUW_ASSERT(input.get_partial_shape()[2].is_static());
             new_shape = ov::PartialShape({1, input_size, input.get_partial_shape()[2]});
+        } else if (input_name.find("deepstack_visual_embeds") != std::string::npos) {
+            // NB: VLMs case (Qwen3-VL "DeepStack"). The parameter has shape
+            // [num_deepstack_layers, num_visual_tokens, EMB_SIZE]: dim 0 selects one of
+            // several intermediate ViT feature levels (consumed by Gather/select ops),
+            // it is NOT a batch dimension. Both num_deepstack_layers and num_visual_tokens
+            // are dynamic in the source model. The visual-token dim is made static to
+            // input_size, while the layer dim MUST be preserved - collapsing it to 1
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/reshape_to_static.cpp]: I think we can drop "must be preserved" here and write that we are calculating a real number here

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp [hunk 6]
```diff
@@ -836,6 +837,8 @@ ov::npuw::LLMCompiledModel::LLMCompiledModel(const std::shared_ptr<ov::Model>& m
     m_bf16_consts = ov::npuw::s11n::get_bf16_consts(model);
     ov::pass::ConvertPrecision(ov::element::bf16, ov::element::f16).run_on_model(kvcache_model);
 
+    ov::npuw::ReplaceDeepstackScatterWithAdd().run_on_model(kvcache_model);
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:840]: If this is unconditional, do we need to run full validation to check that this transformation doesn't affect other models?

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.hpp [hunk 7]
```diff
@@ -56,25 +56,33 @@ class LLMInferRequest : public ov::npuw::LLMInferBaseRequest {
     void infer_chunked_prefill(ov::SoPtr<ov::ITensor> input_ids,
                                ov::SoPtr<ov::ITensor> attention_mask,
                                ov::SoPtr<ov::ITensor> position_ids,
-                               ov::SoPtr<ov::ITensor> per_layer_inputs);
+                               ov::SoPtr<ov::ITensor> per_layer_inputs,
+                               ov::SoPtr<ov::ITensor> visual_pos_masks,
+                               ov::SoPtr<ov::ITensor> deepstack_visual_embeds);
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.hpp:59]: Someday we would need to introduce `ModelInput` or `Context` here..

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp [hunk 8]
```diff
@@ -89,6 +91,119 @@ std::pair<uint32_t, uint32_t> get_lora_dims_by_name(const std::string& state_nam
...
+    }
+    return count;
+}
+
+// Scatters the compact deepstack_visual_embeds tensor (one row per visual token, in
+// visual-token order) into the right-aligned static destination at the actual
+// visual-token sequence positions described by visual_pos_masks.
+//
+// After the DeepStack gather/scatter cluster is replaced in the graph by a plain
+// dense residual add (see densify_deepstack_visual_embeds in llm_compiled_model.cpp),
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp]: I think that there is no such function `densify_deepstack_visual_embeds()` anymore

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp [hunk 9]
```diff
@@ -89,6 +91,119 @@ std::pair<uint32_t, uint32_t> get_lora_dims_by_name(const std::string& state_nam
...
+    const size_t src_seq = src_shape[1];
+    const size_t emb = src_shape[2];
+    const size_t dst_seq = dst_shape[1];
+    if (dst_shape[0] != num_layers || dst_shape[2] != emb) {
+        return 0;
+    }
+
+    const size_t mask_total = mask->get_size();
+    OPENVINO_ASSERT(mask_total <= dst_seq);
+    const size_t seq_right_pad = dst_seq - mask_total;
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp]: Isn't it left padding actually (left padding == right alignment)?

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp [hunk 10]
```diff
@@ -1013,7 +1184,9 @@ void ov::npuw::LLMInferRequest::infer_generate(ov::SoPtr<ov::ITensor> input_ids,
                                                ov::SoPtr<ov::ITensor> attention_mask,
                                                ov::SoPtr<ov::ITensor> position_ids,
                                                ov::SoPtr<ov::ITensor> token_type_ids,
-                                               ov::SoPtr<ov::ITensor> per_layer_inputs) {
+                                               ov::SoPtr<ov::ITensor> per_layer_inputs,
+                                               ov::SoPtr<ov::ITensor> visual_pos_masks,
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp]: Might be we don't need these inputs if they are unused in the generate stage?

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp [hunk 11]
```diff
@@ -1199,6 +1379,16 @@ void ov::npuw::LLMInferRequest::infer() {
 
     auto token_type_ids = ov::npuw::util::TensorPtr();
 
+    auto visual_pos_masks = ov::npuw::util::TensorPtr();
+    if (auto port = ov::npuw::util::find_port_by_name(inputs, "visual_pos_masks"); port.has_value()) {
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp]: Suggest to add these constants to `LLMInferRequest::layer_names` constants.

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp [hunk 12]
```diff
@@ -1199,6 +1379,16 @@ void ov::npuw::LLMInferRequest::infer() {
...
+        visual_pos_masks = get_tensor(port.value());
+    }
+
+    auto deepstack_visual_embeds = ov::npuw::util::TensorPtr();
+    if (auto port = ov::npuw::util::find_port_by_name(inputs, "deepstack_visual_embeds"); port.has_value()) {
+        deepstack_visual_embeds = get_tensor(port.value());
+    }
+
     if (auto type_ids_port = ov::npuw::util::find_port_by_name(inputs, layer_names::token_type_ids);
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp]: Suggest to move it upper: to `token_type_ids` definition

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp [hunk 13]
```diff
@@ -1265,7 +1461,13 @@ void ov::npuw::LLMInferRequest::infer() {
...
             trim_kvcache_for_speculative_decoding(position_ids);
         }
-        infer_generate(input_ids, attention_mask, position_ids, token_type_ids, per_layer_inputs);
+        infer_generate(input_ids,
+                       attention_mask,
+                       position_ids,
+                       token_type_ids,
+                       per_layer_inputs,
+                       visual_pos_masks,
+                       deepstack_visual_embeds);
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp]: Still thinking that we don't need these parameters to pass into generate

File: src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/replace_deepstack_scatter_with_add.cpp (hunk 3 above)
Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/replace_deepstack_scatter_with_add.cpp:27]: yes? Not sure I get you

File: src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/replace_deepstack_scatter_with_add.cpp (hunk 4 above)
Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/replace_deepstack_scatter_with_add.cpp]: done

File: src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/reshape_to_static.cpp (hunk 5 above)
Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/npuw_transformations/reshape_to_static.cpp]: done

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp (hunk 6 above)
Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:840]: don't think we need to full validation, but we need to run some validation.
Condition is inside pass itself

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.hpp (hunk 7 above)
Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.hpp:59]: What do we say to context? not today

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp (hunk 8 above)
Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp]: fixed

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp (hunk 9 above)
Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp]: fixed

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp (hunk 10 above)
Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp]: removed

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp (hunk 11 above)
Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp]: done

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp (hunk 13 above)
Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp]: done

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp (hunk 12 above)
Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp]: done

**Review by AsyaPronina — APPROVED** [review]
LGTM! Thanks a lot, Alexander!

**Comment by AlexanderKalistratov** [comment]
build_jenkins
